### PR TITLE
[SHELL32] Drive Property Sheet: Run in another thread

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

The drive property sheet was modal, so it used to lock the main thread. This PR will unlock by using another thread.

JIRA issue: [CORE-4217](https://jira.reactos.org/browse/CORE-4217)
JIRA issue: [CORE-11547](https://jira.reactos.org/browse/CORE-11547)

## Proposed changes

In `dll/win32/shell32/dialogs/drive.cpp`:

- Add `DRIVE_PROP_DATA` structure.
- Call `SHCreateThread` function in `SH_ShowDriveProperties` function.
- Clean up the data when they should be released.

## TODO

- [x] Do tests.

## Comparison

BEFORE:
before
FAILED. Not working well.

AFTER:
after
It is working well.

### CODE REVIEW

File: dll/win32/shell32/dialogs/drive.cpp [hunk 1]
```diff
@@ -166,9 +166,20 @@ typedef struct _DRIVE_PROP_PAGE
...
-BOOL
-SH_ShowDriveProperties(WCHAR *pwszDrive, IDataObject *pDataObj)
+struct DRIVE_PROP_DATA
+{
+    LPWSTR pwszDrive;
+    IDataObject *pDataObj;
+};
+
+static DWORD WINAPI
+SH_ShowDrivePropThreadProc(LPVOID pParam)
```
Comment (whindsaks) [inline dll/win32/shell32/dialogs/drive.cpp:176]: This is not a public helper function, it does not need a SH prefix.

File: dll/win32/shell32/dialogs/drive.cpp [hunk 2]
```diff
@@ -234,6 +245,32 @@ SH_ShowDriveProperties(WCHAR *pwszDrive, IDataObject *pDataObj)
...
+        return FALSE;
+
+    // Prepare data for thread
+    DRIVE_PROP_DATA *pData = (DRIVE_PROP_DATA *)SHAlloc(sizeof(*pData));
+    if (!pData)
+    {
+        SHFree(pwszDrive);
+        return FALSE;
+    }
+    *pData = { pwszDrive, pDataObj };
```
Comment (whindsaks) [inline dll/win32/shell32/dialogs/drive.cpp:261]: pDataObj is an object you don't control and shell objects are often STA, you can't just pass them to another thread without COM marshaling.

Comment (katahiromz) [reply dll/win32/shell32/dialogs/drive.cpp:261]: @whindsaks I'll leave it up to you to improve this PR. Feel free to copy and modify my code. You understand shell better than I do.

Comment (whindsaks) [reply dll/win32/shell32/dialogs/drive.cpp:261]: No, incorrect code should not be merged in the first place!

**Comment by katahiromz** [comment]
Incorrect code. Closed.

File: dll/win32/shell32/dialogs/drive.cpp (hunk 2 above)
Comment (HBelusca) [reply dll/win32/shell32/dialogs/drive.cpp:261]: https://learn.microsoft.com/en-us/windows/win32/shell/ctf
At the end of this page there's a remark about problems of manipulating STA objects in a separate thread.
Some related info:
https://devblogs.microsoft.com/oldnewthing/20151020-00/?p=91321
https://devblogs.microsoft.com/oldnewthing/20151021-00/?p=91311

**Comment by katahiromz** [comment]
@whindsaks Marshalling is implemented.

File: dll/win32/shell32/dialogs/drive.cpp (hunk 1 above)
Comment (katahiromz) [reply dll/win32/shell32/dialogs/drive.cpp]: Fixed.

File: dll/win32/shell32/dialogs/drive.cpp [hunk 3]
```diff
@@ -234,6 +248,41 @@ SH_ShowDriveProperties(WCHAR *pwszDrive, IDataObject *pDataObj)
...
+        SHFree(pwszDrive);
+        SHFree(pData);
+        return FALSE;
+    }
+
+    // Run a property sheet in another thread
+    if (SHCreateThread(ShowDrivePropThreadProc, pData, CTF_COINIT, NULL))
+        return TRUE; // Success
+
+    pData->pStream->Release();
```
Comment (HBelusca) [inline dll/win32/shell32/dialogs/drive.cpp:280]:
```suggestion
    if (SUCCEEDED(hr))
    {
        // Run a property sheet in another thread
        if (SHCreateThread(ShowDrivePropThreadProc, pData, CTF_COINIT, NULL))
            return TRUE; // Success

        pData->pStream->Release();
    }
```

File: dll/win32/shell32/dialogs/drive.cpp [hunk 4]
```diff
@@ -166,9 +166,23 @@ typedef struct _DRIVE_PROP_PAGE
     UINT DriveType;
 } DRIVE_PROP_PAGE;
 
-BOOL
-SH_ShowDriveProperties(WCHAR *pwszDrive, IDataObject *pDataObj)
+struct DRIVE_PROP_DATA
+{
+    LPWSTR pwszDrive;
```
Comment (HBelusca) [inline dll/win32/shell32/dialogs/drive.cpp:171]:
```suggestion
    PCWSTR pwszDrive;
```

File: dll/win32/shell32/dialogs/drive.cpp [hunk 5]
```diff
@@ -234,6 +248,41 @@ SH_ShowDriveProperties(WCHAR *pwszDrive, IDataObject *pDataObj)
...
+        return FALSE;
+    }
+
+    // Run a property sheet in another thread
+    if (SHCreateThread(ShowDrivePropThreadProc, pData, CTF_COINIT, NULL))
+        return TRUE; // Success
+
+    pData->pStream->Release();
+    SHFree(pwszDrive);
+    SHFree(pData);
```
Comment (HBelusca) [inline dll/win32/shell32/dialogs/drive.cpp:282]:
```suggestion
    SHFree(pData);
    SHFree(pwszDrive);
```
destruction in reverse order of allocation.

File: dll/win32/shell32/dialogs/drive.cpp [hunk 6]
```diff
@@ -166,9 +166,23 @@ typedef struct _DRIVE_PROP_PAGE
     UINT DriveType;
 } DRIVE_PROP_PAGE;
 
-BOOL
-SH_ShowDriveProperties(WCHAR *pwszDrive, IDataObject *pDataObj)
+struct DRIVE_PROP_DATA
+{
+    PCWSTR pwszDrive;
```
Comment (katahiromz) [inline dll/win32/shell32/dialogs/drive.cpp:171]:
```suggestion
    PWSTR pwszDrive;
```

Comment (HBelusca) [reply dll/win32/shell32/dialogs/drive.cpp]: The string cannot be a const here?

Comment (katahiromz) [reply dll/win32/shell32/dialogs/drive.cpp]: It will be released.

**Review by HBelusca — APPROVED** [review]
looks OK at first sight, but a check by @whindsaks or whoever else knowledgeable about COM marshalling would be great.

**Comment by whindsaks** [comment]
LGTM

**Pull request closed without merging**